File: frontend/src/api/dashboards.ts
import type { Paginated } from "./admin";
import { apiFetch } from "./client";

// ── Public dashboards viewer ──────────────────────────────────────────────────

export interface DashboardSummary {
  id: string;
  name: string;
  description: string;
  chart_count: number;
  auto_refresh: boolean;
  refresh_interval: string | null;
}

export interface ChartView {
  id: string;
  title: string;
  description: string;
  chart_type: string;
  custom_d3_code: string;
  width: number;
  height: number;
  position: number;
  last_refreshed_at: string | null;
}

export interface DashboardDetail {
  id: string;
  name: string;
  description: string;
  auto_refresh: boolean;
  refresh_interval: string | null;
  charts: ChartView[];
}

export interface ChartData {
  columns: string[];
  data: Record<string, unknown>[];
  cached?: boolean;
  refreshed_at?: string | null;
  error?: string;
}

export function getDashboards(): Promise<{ dashboards: DashboardSummary[] }> {
  return apiFetch("/api/v1/dashboards/");
}
export function getDashboard(id: string): Promise<DashboardDetail> {
  return apiFetch(`/api/v1/dashboards/${id}/`);
}
export function getChartData(id: string, refresh = false): Promise<ChartData> {
  return apiFetch(`/api/v1/charts/${id}/data/${refresh ? "?refresh=1" : ""}`);
}

// ── Admin: dashboards ─────────────────────────────────────────────────────────

export interface AdminDashboard {
  id: string;
  name: string;
  description: string;
  is_active: boolean;
  auto_refresh: boolean;
  refresh_interval: string | null;
  allowed_roles: string[];
  chart_count: number;
  created_at: string;
  updated_at: string;
}

export interface AdminDashboardInput {
  name?: string;
  description?: string;
  is_active?: boolean;
  auto_refresh?: boolean;
  refresh_interval?: string | null;
  allowed_roles?: string[];
}

const DASH_BASE = "/api/v1/admin/dashboards/";

export function listDashboards(): Promise<Paginated<AdminDashboard>> {
  return apiFetch(DASH_BASE);
}
export function getAdminDashboard(id: string): Promise<AdminDashboard> {
  return apiFetch(`${DASH_BASE}${id}/`);
}
export function createDashboard(data: AdminDashboardInput): Promise<AdminDashboard> {
  return apiFetch(DASH_BASE, { method: "POST", body: JSON.stringify(data) });
}
export function updateDashboard(id: string, data: AdminDashboardInput): Promise<AdminDashboard> {
  return apiFetch(`${DASH_BASE}${id}/`, { method: "PATCH", body: JSON.stringify(data) });
}
export function deleteDashboard(id: string): Promise<void> {
  return apiFetch(`${DASH_BASE}${id}/`, { method: "DELETE" });
}

// ── Admin: charts ─────────────────────────────────────────────────────────────

export interface AdminChart {
  id: string;
  dashboard: string;
  database: string;
  database_name: string;
  title: string;
  description: string;
  sql_query: string;
  custom_d3_code: string;
  chart_type: string;
  width: number;
  height: number;
  position: number;
  is_active: boolean;
  last_error: string;
  last_refreshed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface AdminChartInput {
  dashboard?: string;
  database?: string;
  title?: string;
  description?: string;
  sql_query?: string;
  custom_d3_code?: string;
  width?: number;
  height?: number;
  position?: number;
  is_active?: boolean;
}

export interface PreviewResult {
  columns: string[];
  data: Record<string, unknown>[];
  error?: string;
}

const CHART_BASE = "/api/v1/admin/charts/";

export function listCharts(dashboardId: string): Promise<Paginated<AdminChart>> {
  return apiFetch(`${CHART_BASE}?dashboard=${encodeURIComponent(dashboardId)}`);
}
export function getChart(id: string): Promise<AdminChart> {
  return apiFetch(`${CHART_BASE}${id}/`);
}
export function createChart(data: AdminChartInput): Promise<AdminChart> {
  return apiFetch(CHART_BASE, { method: "POST", body: JSON.stringify(data) });
}
export function updateChart(id: string, data: AdminChartInput): Promise<AdminChart> {
  return apiFetch(`${CHART_BASE}${id}/`, { method: "PATCH", body: JSON.stringify(data) });
}
export function deleteChart(id: string): Promise<void> {
  return apiFetch(`${CHART_BASE}${id}/`, { method: "DELETE" });
}
export function previewChart(database: string, sql_query: string): Promise<PreviewResult> {
  return apiFetch(`${CHART_BASE}preview/`, {
    method: "POST",
    body: JSON.stringify({ database, sql_query }),
  });
}
export function getAdminChartData(id: string, refresh = false): Promise<ChartData> {
  return apiFetch(`${CHART_BASE}${id}/data/${refresh ? "?refresh=1" : ""}`);
}

// ── Admin: AI dashboard generation ────────────────────────────────────────────

export interface DashboardGenOptions {
  databases: { id: string; name: string }[];
  doc_sources: { id: string; name: string }[];
  codebases: { id: string; name: string }[];
  agents: { id: string; name: string; is_active: boolean }[];
  dashboard_types: string[];
}

export interface DashboardGenerateRequest {
  dashboard_name: string;
  dashboard_type: string;
  prompt_override?: string;
  agent: string;
  source_db?: string[];
  source_doc?: string[];
  source_codebase?: string[];
}

export interface DashboardGenStatus {
  id: string;
  status: "running" | "success" | "partial" | "failed";
  execution_time_ms: number | null;
  dashboard_id?: string | null;
  charts_created?: number;
  agent_output?: string;
  error?: string;
  dashboard_name?: string;
}

export interface ChartGenLog {
  id: string;
  status: string;
  dashboard_name: string;
  charts_created: number | null;
  execution_time_ms: number | null;
  error_message: string;
  user: string | null;
  agent: string | null;
  started_at: string;
  completed_at: string | null;
}

export interface ChartGenLogDetail extends ChartGenLog {
  errors: string[];
  source_databases: string[];
  source_docs: string[];
  prompt_used: string;
  agent_output: string;
}

export function getDashboardGenerateOptions(): Promise<DashboardGenOptions> {
  return apiFetch(`${DASH_BASE}generate-options/`);
}
export function startDashboardGenerate(
  data: DashboardGenerateRequest
): Promise<{ log_id: string }> {
  return apiFetch(`${DASH_BASE}generate/`, { method: "POST", body: JSON.stringify(data) });
}

const CHARTGEN_LOG_BASE = "/api/v1/admin/chartgen-logs/";

export function getDashboardGenStatus(logId: string): Promise<DashboardGenStatus> {
  return apiFetch(`${CHARTGEN_LOG_BASE}${logId}/status/`);
}
export function listChartGenLogs(): Promise<Paginated<ChartGenLog>> {
  return apiFetch(CHARTGEN_LOG_BASE);
}
export function getChartGenLog(id: string): Promise<ChartGenLogDetail> {
  return apiFetch(`${CHARTGEN_LOG_BASE}${id}/`);
}

File: backend/api/v1/admin/dashboards.py
"""Dashboard + chart admin API: CRUD, ad-hoc preview, chart data, AI generation.

Ports ``management/views/dashboard.py``. Charts are custom-d3 (the model's other
``chart_type`` values are legacy); SQL is validated read-only via the serializer.
AI generation is delegated to ``engine.engines.chartgen`` and tracked by
``ChartGenerationLog``, polled through the ``status`` action.
"""

from __future__ import annotations

from typing import Any, cast

from django.contrib.auth.models import User
from django.db.models import Count, QuerySet
from engine.engines import chartgen
from engine.engines.charts import chart_data, preview_query
from engine.models import (
    AgentConfiguration,
    Chart,
    ChartGenerationLog,
    Codebase,
    Dashboard,
    DatabaseConnection,
    DocumentationSource,
)
from engine.prompts import DASHBOARD_TEMPLATES
from rest_framework import serializers, status, viewsets
from rest_framework.decorators import action
from rest_framework.request import Request
from rest_framework.response import Response

from api.permissions import IsStaffUser
from api.serializer_meta import SerializerMeta


def _resolve_names(model: Any, field: str, ids: list[str]) -> list[str]:
    return list(model.objects.filter(pk__in=ids, is_active=True).values_list(field, flat=True))


class DashboardSerializer(serializers.ModelSerializer[Dashboard]):
    chart_count = serializers.IntegerField(read_only=True)

    class Meta(SerializerMeta):
        model = Dashboard
        fields = [
            "id",
            "name",
            "description",
            "is_active",
            "auto_refresh",
            "refresh_interval",
            "allowed_roles",
            "chart_count",
            "created_at",
            "updated_at",
        ]
        read_only_fields = ["id", "chart_count", "created_at", "updated_at"]


class DashboardViewSet(viewsets.ModelViewSet[Dashboard]):
    """Staff CRUD for dashboards, plus AI generation + its options."""

    permission_classes = [IsStaffUser]
    serializer_class = DashboardSerializer

    def get_queryset(self) -> QuerySet[Dashboard]:
        return Dashboard.objects.annotate(chart_count=Count("charts")).order_by("name")

    def perform_create(self, serializer: Any) -> None:
        serializer.save(created_by=self.request.user)

    @action(detail=False, methods=["get"], url_path="generate-options")
    def generate_options(self, request: Request) -> Response:
        return Response(
            {
                "databases": [
                    {"id": d.pk, "name": d.name}
                    for d in DatabaseConnection.objects.filter(is_active=True)
                ],
                "doc_sources": [
                    {"id": s.pk, "name": s.folder_name}
                    for s in DocumentationSource.objects.filter(is_active=True)
                ],
                "codebases": [
                    {"id": c.pk, "name": c.name} for c in Codebase.objects.filter(is_active=True)
                ],
                "agents": [
                    {"id": a.pk, "name": a.name, "is_active": a.is_active}
                    for a in AgentConfiguration.objects.all()
                ],
                "dashboard_types": sorted(DASHBOARD_TEMPLATES),
            }
        )

    @action(detail=False, methods=["post"])
    def generate(self, request: Request) -> Response:
        data = request.data
        dashboard_name = (data.get("dashboard_name") or "").strip()
        agent_id = data.get("agent")
        if not all([dashboard_name, agent_id]):
            return Response(
                {"detail": "Missing required fields."}, status=status.HTTP_400_BAD_REQUEST
            )
        if Dashboard.objects.filter(name=dashboard_name).exists():
            return Response(
                {"detail": f"A dashboard named '{dashboard_name}' already exists."},
                status=status.HTTP_400_BAD_REQUEST,
            )
        agent_config = AgentConfiguration.objects.filter(pk=agent_id).first()
        if agent_config is None:
            return Response({"detail": "Agent not found."}, status=status.HTTP_404_NOT_FOUND)

        log = chartgen.start_generation(
            user=cast(User, request.user),
            agent_config=agent_config,
            dashboard_name=dashboard_name,
            dashboard_type=data.get("dashboard_type", "overview"),
            prompt_override=data.get("prompt_override", ""),
            db_names=_resolve_names(DatabaseConnection, "name", data.get("source_db", [])),
            doc_names=_resolve_names(
                DocumentationSource, "folder_name", data.get("source_doc", [])
            ),
            codebase_names=_resolve_names(Codebase, "name", data.get("source_codebase", [])),
        )
        return Response({"log_id": log.pk}, status=status.HTTP_202_ACCEPTED)


class ChartSerializer(serializers.ModelSerializer[Chart]):
    database_name = serializers.CharField(source="database.name", read_only=True)

    class Meta(SerializerMeta):
        model = Chart
        fields = [
            "id",
            "dashboard",
            "database",
            "database_name",
            "title",
            "description",
            "sql_query",
            "custom_d3_code",
            "chart_type",
            "width",
            "height",
            "position",
            "is_active",
            "last_error",
            "last_refreshed_at",
            "created_at",
            "updated_at",
        ]
        read_only_fields = [
            "id",
            "database_name",
            "last_error",
            "last_refreshed_at",
            "created_at",
            "updated_at",
        ]

    def validate(self, attrs: dict[str, Any]) -> dict[str, Any]:
        from engine.engines.report_engine import validate_sql

        sql = attrs.get("sql_query", getattr(self.instance, "sql_query", ""))
        db = attrs.get("database", getattr(self.instance, "database", None))
        if sql:
            error = validate_sql(sql, engine=db.engine if db else None)
            if error:
                raise serializers.ValidationError({"sql_query": error})
        return attrs

    def create(self, validated_data: dict[str, Any]) -> Chart:
        validated_data.setdefault("chart_type", "custom")
        return super().create(validated_data)


class ChartViewSet(viewsets.ModelViewSet[Chart]):
    """Staff CRUD for charts (filter by ``?dashboard=``), plus preview + data."""

    permission_classes = [IsStaffUser]
    serializer_class = ChartSerializer

    def get_queryset(self) -> QuerySet[Chart]:
        qs = Chart.objects.select_related("database", "dashboard")
        dashboard_id = self.request.query_params.get("dashboard")
        if dashboard_id:
            qs = qs.filter(dashboard_id=dashboard_id)
        return qs.order_by("position", "title")

    @action(detail=False, methods=["post"])
    def preview(self, request: Request) -> Response:
        """Run an ad-hoc ``{database, sql_query}`` query for the editor preview."""
        database_id = request.data.get("database")
        sql = (request.data.get("sql_query") or "").strip()
        if not database_id:
            return Response({"error": "database is required."}, status=status.HTTP_400_BAD_REQUEST)
        if not sql:
            return Response({"error": "sql_query is required."}, status=status.HTTP_400_BAD_REQUEST)
        db_conn = DatabaseConnection.objects.filter(pk=database_id).first()
        if db_conn is None:
            return Response({"error": "Database not found."}, status=status.HTTP_404_NOT_FOUND)
        try:
            return Response(preview_query(db_conn, sql))
        except ValueError as exc:
            return Response({"error": str(exc)}, status=status.HTTP_400_BAD_REQUEST)
        except Exception as exc:
            return Response({"error": str(exc)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)

    @action(detail=True, methods=["get"])
    def data(self, request: Request, pk: str | None = None) -> Response:
        chart = self.get_object()
        force_refresh = request.query_params.get("refresh") == "1"
        try:
            return Response(chart_data(chart, force_refresh))
        except Exception as exc:
            return Response({"error": str(exc)}, status=status.HTTP_500_INTERNAL_SERVER_ERROR)


class ChartGenerationLogSerializer(serializers.ModelSerializer[ChartGenerationLog]):
    user = serializers.CharField(source="user.username", read_only=True, default=None)
    agent = serializers.CharField(source="agent.name", read_only=True, default=None)

    class Meta(SerializerMeta):
        model = ChartGenerationLog
        fields = [
            "id",
            "status",
            "dashboard_name",
            "charts_created",
            "execution_time_ms",
            "error_message",
            "user",
            "agent",
            "started_at",
            "completed_at",
        ]
        read_only_fields = fields


class ChartGenerationLogDetailSerializer(serializers.ModelSerializer[ChartGenerationLog]):
    """Full record for the detail page — adds the prompt, agent output, and sources."""

    user = serializers.CharField(source="user.username", read_only=True, default=None)
    agent = serializers.CharField(source="agent.name", read_only=True, default=None)

    class Meta(SerializerMeta):
        model = ChartGenerationLog
        fields = [
            "id",
            "status",
            "dashboard_name",
            "charts_created",
            "execution_time_ms",
            "error_message",
            "errors",
            "source_databases",
            "source_docs",
            "prompt_used",
            "agent_output",
            "user",
            "agent",
            "started_at",
            "completed_at",
        ]
        read_only_fields = fields


class ChartGenerationLogViewSet(viewsets.ReadOnlyModelViewSet[ChartGenerationLog]):
    """Read-only dashboard-generation history + a ``status`` payload for polling.

    ``list`` uses the lean serializer; ``retrieve`` returns the full record
    (prompt + agent output can be large, so they're detail-only).
    """

    permission_classes = [IsStaffUser]
    serializer_class = ChartGenerationLogSerializer
    queryset = ChartGenerationLog.objects.select_related("user", "agent").all()

    def get_serializer_class(self) -> type[serializers.BaseSerializer[ChartGenerationLog]]:
        if self.action == "retrieve":
            return ChartGenerationLogDetailSerializer
        return ChartGenerationLogSerializer

    @action(detail=True, methods=["get"])
    def status(self, request: Request, pk: str | None = None) -> Response:
        return Response(chartgen.status_payload(self.get_object()))
